dnscrypt-proxy/coldstart.go

```diff
@@ -95,6 +96,12 @@ func handleColdStartClient(clientPc *net.UDPConn, cancelChannel chan struct{}, i
 	}
 	if err != nil {
 		dlog.Warn(err)
+		select {
+		case <-cancelChannel:
+			cancelChannel <- struct{}{}
+		default:
+			cancelChannel = make(chan struct{}, 1)
```

Comment:
Is that line correct? We create a channel, but we immediately return so nothing's gonna be pushed to it.

Reply:
(yes, that's a late comment, but I stumbled upon that right before tagging a release)

Reply:
It is for these unbuffered channels:
https://github.com/DNSCrypt/dnscrypt-proxy/blob/3d641b758a98f712ed95e4ffc89eff327c139679/dnscrypt-proxy/coldstart.go#L21-L26
The workflow is:
1. `Stop` tries to send a signal to each `handleColdStartClient` to exit. It blocks until a `select` in `handleColdStartClient` receives from the channel and then delivers a new value to equal the send/receive balance.
2. The 2 `return true` of `handleColdStartClient` will end rerunning `handleColdStartClient`. The 2nd `select` is in case of the `Stop` sends signals after the 1st one. If that is between the 2 `select`s, we relieve the channel blocking. If signal is still not sent, we change the channel to a buffered channel, which doesn't block. The 1st `return true` is from the 1st `select`, and has had a balanced consumed channel.

A [demo](https://gist.github.com/lifenjoiner/6de299af7b8ef5a27ac6c8ea1160c3e4) may help on this unbuffered channel and select behaviors topic.

Reply:
I made it complicated. I understand it better now.
A buffered channel is good enough. It is non-blocking for sending, for `default` case of `select`, and for `select` on closed channel too.
Optimized with https://github.com/DNSCrypt/dnscrypt-proxy/pull/2151.

Reply:
Leave a channel open is OK: https://stackoverflow.com/questions/8593645/is-it-ok-to-leave-a-channel-open, but blocked is not a good choice.